zml/tokenizer.zig

```diff
@@ -120,16 +122,69 @@ pub const Tokenizer = struct {
         return self.token_lookup.get(str);
     }
 
+    fn toHex(dst: []u8, src: []const u8, comptime case: std.fmt.Case) ![]u8 {
+        return std.fmt.bufPrint(dst, "<0x{s}>", .{switch (case) {
+            .lower => std.fmt.fmtSliceHexLower(src),
+            else => std.fmt.fmtSliceHexUpper(src),
+        }});
+    }
+
+    pub fn preprocess(self: *Tokenizer, allocator: std.mem.Allocator, raw: []const u8) ![]const u8 {
+        var out = std.ArrayList(u8).init(allocator);
+
+        var utf8 = (try std.unicode.Utf8View.init(raw)).iterator();
+        while (utf8.nextCodepointSlice()) |codepoint| {
+            // do not handle spaces (normalizer handles)
+            if (std.mem.eql(u8, codepoint, " ")) {
+                try out.appendSlice(codepoint);
+                continue;
+            }
+            // If found in `Tokenizer.lookup`, append it
+            if (self.lookup(codepoint)) |_| {
+                try out.appendSlice(codepoint);
+                continue;
+            }
+
+            const arena = self.arena_state.allocator();
+
+            // If uppercase hex representation is found, append
+            // (e.g. Huggingface models sometimes converts `\n` to `<0x0A>`).
+            var upper_buffer: [100]u8 = undefined;
+            const upper_hex = try toHex(&upper_buffer, codepoint, .upper);
+            if (self.lookup(upper_hex)) |id| {
+                try self.mapped_tokens.put(arena, codepoint, id);
+                try out.appendSlice(upper_hex);
+                continue;
+            }
+
+            // If lowercase hex representation is found, append
+            // (e.g. Huggingface models sometimes converts `\n` to `<0x0A>`).
+            var lower_buffer: [100]u8 = undefined;
+            const lower_hex = try toHex(&lower_buffer, codepoint, .lower);
+            if (self.lookup(lower_hex)) |id| {
+                try self.mapped_tokens.put(arena, codepoint, id);
```

Comment:
it's not the right time to do this.
I think that when creating the tokenizer we should convert `<0x0A>` to EOL.
then you wouldn't need to preprocess them and post-process them too.